vm-virtio/src/vhost_user/net.rs

```diff
@@ -0,0 +1,315 @@
+// Copyright 2019 Intel Corporation. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0 AND BSD-3-Clause
+
+use libc;
+use libc::EFD_NONBLOCK;
+use std::cmp;
+use std::io::Write;
+use std::os::unix::io::AsRawFd;
+use std::sync::Arc;
+use std::thread;
+use std::vec::Vec;
+
+use crate::VirtioInterrupt;
+use net_util::{MacAddr, MAC_ADDR_LEN};
+
+use vm_memory::{Address, Error as MmapError, GuestMemory, GuestMemoryMmap, GuestMemoryRegion};
+use vmm_sys_util::eventfd::EventFd;
+
+use super::super::{ActivateError, ActivateResult, Queue, VirtioDevice, VirtioDeviceType};
+use super::handler::VhostUserEpollHandler;
+use super::{Error, Result};
+use vhost_rs::vhost_user::{Master, VhostUserMaster};
+use vhost_rs::{VhostBackend, VhostUserMemoryRegionInfo, VringConfigData};
+use virtio_bindings::virtio_net;
+
+const VIRTIO_F_EVENT_IDX: ::std::os::raw::c_uint = 29;
+const VIRTIO_F_NOTIFY_ON_EMPTY: ::std::os::raw::c_uint = 24;
+const VIRTIO_F_VERSION_1: ::std::os::raw::c_uint = 32;
+const VHOST_USER_F_PROTOCOL_FEATURES: ::std::os::raw::c_uint = 30;
+
+pub struct Net {
+    vhost_user_net: Master,
+    kill_evt: EventFd,
+    avail_features: u64,
+    acked_features: u64,
+    backend_features: u64,
+    config_space: Vec<u8>,
+    queue_sizes: Vec<u16>,
+}
+
+impl Net {
+    /// Create a new vhost-user-net device
+    pub fn new(
+        mac_addr: MacAddr,
+        path: &str,
+        num_queues_pairs: usize,
+        queue_size: u16,
+    ) -> Result<Net> {
+        let num_queues = 2 * num_queues_pairs;
+        let mut vhost_user_net =
+            Master::connect(path, num_queues as u64).map_err(Error::VhostUserCreateMaster)?;
+
+        let kill_evt = EventFd::new(EFD_NONBLOCK).map_err(Error::CreateKillEventFd)?;
+
+        let mut avail_features = 1 << virtio_net::VIRTIO_NET_F_GUEST_CSUM
+            | 1 << virtio_net::VIRTIO_NET_F_CSUM
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_TSO4
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_TSO6
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_ECN
+            | 1 << virtio_net::VIRTIO_NET_F_GUEST_UFO
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_TSO4
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_TSO6
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_ECN
+            | 1 << virtio_net::VIRTIO_NET_F_HOST_UFO
+            | 1 << virtio_net::VIRTIO_NET_F_MRG_RXBUF
+            | 1 << VIRTIO_F_NOTIFY_ON_EMPTY
+            | 1 << VIRTIO_F_VERSION_1
+            | 1 << VIRTIO_F_EVENT_IDX;
+
+        // Get features from backend, do negotiation to get a feature collection which
+        // both VMM and backend can support.
+        let backend_features = vhost_user_net.get_features().unwrap();
```

Comment:
The whole `VIRTIO_FEATURES` and `PROTOCOL_FEATURES` negotiation should be factorized between vhost-user implementations. Both virtio-fs and vhost-user-net should rely on the same function to perform features negotiation.

Reply:
@sboeuf it seems that @bjzhjing implementation is generic enough for virtio-fs to use it. Here `vhost_user_net` is a vu master, which implements the `VhostBackend` trait. And that's where the generic feature negociation implementation lives.